libcob/fileio.c

```diff
@@ -1441,88 +1445,118 @@ cob_chk_file_env (cob_file *f, const char *src)
 	const char	*t;
 	unsigned int	 i;
 
+	/* GC-sanity rule: no environment handling if src starts with period */
+	if (*src == '.') {
+		return NULL;
+	}
+
+	/* no mapping if filename begins with a slash [externally checked], hypen or digits
+	   (taken from "Programmer's Guide to File Handling, Chapter 2: File Naming") */
+	switch (*file_open_name) {
+	case '-':
+	case '0':
+	case '1':
+	case '2':
+	case '3':
+	case '4':
+	case '5':
+	case '6':
+	case '7':
+	case '8':
+	case '9':
+		return NULL;
+	default:
+		break;
+	}
+
+	q = cob_strdup (src);
+	s = q;
 	if (file_setptr->cob_env_mangle) {
-		q = cob_strdup (src);
-		s = q;
 		for (i = 0; s[i] != 0; ++i) {
 			if (!isalnum ((int)s[i])) {
 				s[i] = '_';
 			}
 		}
 	} else {
-		q = NULL;
-		s = (char *)src;
-	}
-
-	if ((file_open_io_env = cob_get_env ("IO_OPTIONS", NULL)) != NULL) {
-		cob_set_file_format (f, file_open_io_env, 1);	/* Set initial defaults */
-	}
-	if (f->organization == COB_ORG_INDEXED) {
-		t = "IX";
-	} else if (f->organization == COB_ORG_SEQUENTIAL) {
-		t = "SQ";
-	} else if (f->organization == COB_ORG_LINE_SEQUENTIAL) {
-		if((f->flag_line_adv & COB_LINE_ADVANCE))
-			t = "LA";
-		else
-			t = "LS";
-	} else if (f->organization == COB_ORG_RELATIVE) {
-		t = "RL";
-	} else {
-		t = "IO";
-	}
-	snprintf (file_open_env, (size_t)COB_FILE_MAX, "%s_OPTIONS", t);
-	if ((file_open_io_env = cob_get_env (file_open_env, NULL)) == NULL) {
-		snprintf (file_open_env, (size_t)COB_FILE_MAX, "%s_options", t);
-		file_open_env[0] = (char)tolower(file_open_env[0]);
-		file_open_env[1] = (char)tolower(file_open_env[1]);
-		file_open_io_env = cob_get_env (file_open_env, NULL);
-	}
-	if (file_open_io_env != NULL) {
-		cob_set_file_format (f, file_open_io_env, 1);	/* Defaults for file type */
+		for (i = 0; s[i] != 0; ++i) {
+			if (s[i] == '.') {
+				s[i] = '_';
+			}
+		}
 	}
 
-	/* Check for IO_filename with file specific options */
-	file_open_io_env = NULL;
-	snprintf (file_open_env, (size_t)COB_FILE_MAX, "%s%s", "IO_", s);
-	if ((file_open_io_env = cob_get_env (file_open_env, NULL)) == NULL) {
-		snprintf (file_open_env, (size_t)COB_FILE_MAX, "%s%s", "io_", s);
+	if (f != NULL) {
```

Comment:
When/why should `f` be null here?

Reply:
Because some functions (`open_cbl_file`, `cob_sys_delete_file`, ...) were "improved" to perform file mapping in GC3 (rev 3944), by calling the `cob_chk_file_mapping` function, which does not take a `cob_file` argument in GC3 but does in GC4, and that function in turn calls `cob_chk_file_env`. Since these functions  (`open_cbl_file`, `cob_sys_delete_file`, ...) do not use a `cob_file` object, I resorted to passing NULL and coping with that...

Reply:
Is this okay for you @GitMensch ?